Src/AmrCore/AMReX_ErrorList.cpp

```diff
@@ -228,4 +228,196 @@ operator << (std::ostream&    os,
+#else
+      auto ay = amrex::Math::abs(dat(i,j+1,k) - dat(i,j,k));
+      ay = amrex::max(ay,amrex::Math::abs(dat(i,j,k) - dat(i,j-1,k)));
+#if AMREX_SPACEDIM > 2
+      auto az = amrex::Math::abs(dat(i,j,k+1) - dat(i,j,k));
+      az = amrex::max(az,amrex::Math::abs(dat(i,j,k) - dat(i,j,k-1)));
+#endif
+      if (amrex::max(D_DECL(ax,ay,az)) >= threshold) {
+        tag(i,j,k) = tagval;
+      }
+#endif
+    });
+  }
+
+  int
+  AMRErrorTag::SetNGrow () const noexcept
+  {
+    AMREX_ALWAYS_ASSERT_WITH_MESSAGE(m_test != USER, "Do not call SetNGrow with USER test");
+    static std::map<TEST,int> ng = { {GRAD,1}, {LESS,0}, {GREATER,0}, {VORT,0}, {BOX,0} };
+    return ng[m_test];
+  }
+  
+  static
+  void
+  AMRErrorTag_LESS(const Box&                bx,
+                   Array4<const Real> const& dat,
+                   Array4<char> const&       tag,
+                   Real                      threshold,
+                   int                       tagval)
+  {
+    amrex::ParallelFor(bx,
+    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+    {
+      if (dat(i,j,k) <= threshold) {
+        tag(i,j,k) = tagval;
+      }
+    });
+  }
+
+  static
+  void
+  AMRErrorTag_GREATER(const Box&                bx,
+                      Array4<const Real> const& dat,
+                      Array4<char> const&       tag,
+                      Real                      threshold,
+                      int                       tagval)
+  {
+    amrex::ParallelFor(bx,
+    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+    {
+      if (dat(i,j,k) >= threshold) {
+        tag(i,j,k) = tagval;
+      }
+    });
+  }
+
+  static
+  void
+  AMRErrorTag_BOX(const Box&          bx,
+                  Array4<char> const& tag,
+                  const RealBox&      tag_rb,
+                  const Geometry&     geom,
+                  int                 tagval)
+  {
+    const Real* dx = geom.CellSize();
+    const Real* plo = geom.ProbLo();
+    class RealBox trb(bx,dx,plo);
+    if (tag_rb.intersects(trb))
+    {
+      amrex::ParallelFor(bx,
+      [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+      {
+        GpuArray<Real,AMREX_SPACEDIM> pt = {D_DECL(plo[0]+(i+0.5)*dx[0],
+                                                   plo[1]+(j+0.5)*dx[1],
+                                                   plo[2]+(k+0.5)*dx[2])};
+        if (tag_rb.contains(pt.data())) {
+          tag(i,j,k) = tagval;
+        }
+      });
+    }
+  }
+
+  static
+  void
+  AMRErrorTag_VORT(const Box&                bx,
+                   Array4<const Real> const& dat,
+                   Array4<char> const&       tag,
+                   int                       level,
+                   Real                      threshold,
+                   int                       tagval)
+  {
+    amrex::ParallelFor(bx,
+    [=] AMREX_GPU_HOST_DEVICE (int i, int j, int k) noexcept
+    {
+      if (dat(i,j,k) >= threshold * std::pow(2,level)) {
```

Comment:
For performance, we should compute `threshold * std::pow(2,level)` outside `ParallelFor`.

Reply:
Will do.